docs/design/storagenode-downtime-tracking.md

```diff
@@ -0,0 +1,202 @@
+    seconds integer,                                                    -- Measured number of seconds being offline
+    CONSTRAINT nodes_offline_time__pk PRIMARY KEY(node_id, tracked_at)
+)
+```
+
+The current Satellite database has the table `nodes`. For the offline time calculation we use the following columns:
+
+- `last_contact_success`
+- `last_contact_failure`
+- `total_uptime_count`
+- `uptime_success_count`
+
+### Detecting offline nodes
+
+An independent process runs, in a configurable interval, of time the following query:
+
+```sql
+SELECT
+FROM nodes
+WHERE
+    last_contact_success < (now()  - 1h) AND
+    last_contact_success > last_contact_failure AND
+    disqualified IS NULL
+ORDER BY
+    last_contact_success ASC
+LIMIT 1
+```
+
+The process runs the query repeatable until it doesn't return records then it ends and will run again in the next interval. The query only gets one record at the time for not holding a record set during the process.
+
+For each node, the Satellite performs an _uptime check_.
+
+* On success
+    ```sql
+        UPDATE nodes
+        SET
+            last_contact_success = now(),
+            uptime_success_count = uptime_success_count + 1,
+            total_uptime_count = total_uptime_count +1
+        WHERE
+            id = ?
+    ```
+* On failure:
+
+  It calculates the number of offline seconds. We consider that the Storage Node must contact the Satellite every hour.
+
+  Because we don't know exactly at what point it has got offline, we dismiss the hour between the last contact success and the supposed next one.
+
+  ```
+  num_seconds_offline =  seconds(from: last_contact_success, to: now() - 1h)
+  ```
+
+  ```sql
+  INSERT INTO nodes_offline_time (node_id, tracked_time, seconds)
+  VALUES (<<id>>, now(), <<num_seconds_offline>>)
+  ```
+
+  ```sql
+  UPDATE nodes
+  SET
+    last_contact_failure = now(),
+    total_uptime_count = total_uptime_count +1
+  WHERE
+    id = ?
+   ```
+
+### Estimating offline time
+
+An independent process has the following configurable parameters:
+
+- Sleep time.
+- Number of nodes to check.
+
+The process enters in a loop until it finishes, running the following query:
+
+```sql
+SELECT
+FROM nodes
+WHERE
+    last_contact_success < last_contact_failure AND
+    disqualified IS NULL
+ORDER BY
+    last_contact_failure ASC
+LIMIT N
+```
+
+The query uses the number of nodes parameter for limiting the records.
+It checks the configured number of nodes, then will sleep the configured amount of time and start again.
+
+For each node performs an _uptime check_.
+
+* On success:
+  ```sql
+  UPDATE nodes
+  SET
+    last_contact_success = now(),
+    uptime_success_count = uptime_success_count + 1,
+    total_uptime_count = total_uptime_count +1
+  WHERE
+    id = ?
+  ```
+* On failure:
+
+  Calculate the number of seconds offline from now and the last contact failure.
+  ```
+  num_seconds_offline =  seconds(from: last_contact_failure, to: now())
```

Comment:
won't this only work properly if the last contact failure also updated the `nodes_offline_time` table? We will have to make sure we do not update `last_contact_success` or `last_contact_failure` without also updating the new table.

Reply:
That's a very good point @mobyvb 
It's a case that I didn't contemplate.

I've been thinking in which case we wouldn't track some offline period of time and I drew a diagram for being clear about it and, hopefully, to show the cases to others easily.

![offline tracking-Offline tracking with other audit updates](https://user-images.githubusercontent.com/1731633/63583808-01f64e80-c59c-11e9-8212-89d0b0b9b27a.png)

We have 2 choices:

1. Assume that we will lose some amount of offline tracking applying the premise commented at the beginning of the design section "Roughly estimate how long they are offline".
2. Do something about it.

### Choice 1 

M understanding that there won't be that many time on what we would lose the tracking of that offline time.

If I'm correct, _audit_ is mostly the procedure which updates the _uptime failure_ in an unregular and "indeterministic" way.

Audit service picks a random stripe and gets the pointer to the segment for downloading erasure shares. During that process, it can detect the Storage Nodes which are offline and update the `last_contact_failure`.

Because of its randomness I assuming that there won't be that many cases where `last_contact_failure` is updated before the _estimate offline time process_ select the node to verify if it's still offline.

What I dislike of this solution is that there is no clear bounding what it's in charge of updating and accessing to a certain part of the database, however, that isn't unusual in a monolith system.

### Choice 2

I've spent some time thinking about how we could track offline time.

Any solution, using the database (adding more columns, tables, queries, etc.) that came to the top of my head didn't work, while the time could be tracked, I found that there were some cases that we would track for some SN more offline track that the one that we know for sure that the SN is offline. That's unadmissible.

I've found a solution. `last_contact_failure` is only updated by the _SN downtime tracking service_. Any other process/service has to notify this service when it finds an SN which is offline.

An alternate solution is to rearhcitecture the system. Have an isolated service which will be in charge of calculating the node reputation, so it would execute the audits, uptimes checks, etc. It isn't a solution itself, mostly a way to rethink a part of the system which is related. Obviously, this requires much more work and it may be part of work being done to split the satellite in different independent services.


I'll present this on today's architecture meeting.

Reply:
> I've found a solution. last_contact_failure is only updated by the SN downtime tracking service. Any other process/service has to notify this service when it finds an SN which is offline.

I think the issue with this is it could have an unexpected performance impact. And part of the reason for redefining how we do uptime tracking is to try not to do as much db work for every little thing that might indirectly affect uptime. The idea of interconnecting any service that might detect that a node is offline with this downtime tracker seems intimidating to me (for a first implementation, at least).

For that reason and the other reasons you provided, I think it is better to go with choice 1, and disregard uptime data that is outside the scope of the processes you have defined in this document. It should still be enough data for us to provide nodes with sufficient proof for disqualification, and will be less complicated than the alternative.

Reply:
> I think the issue with this is it could have an unexpected performance impact.

Yes, it should be slower because currently, the parts update the DB directly, after this, they will have to contact to this new service/process and it will update the DB, hence it adds an indirection.

>  And part of the reason for redefining how we do uptime tracking is to try not to do as much db work for every little thing that might indirectly affect uptime.

I don't think that the DB will be more used, the number of updates that it should receive will be the same, but rather than being done directly by some different parts of the code, they will be "queued" to the new service/process.

> For that reason and the other reasons you provided, I think it is better to go with choice 1, and disregard uptime data that is outside the scope of the processes you have defined in this document. It should still be enough data for us to provide nodes with sufficient proof for disqualification, and will be less complicated than the alternative.

I fully agree with that.

Reply:
> Audit service picks a random stripe and gets the pointer to the segment for downloading erasure shares. During that process, it can detect the Storage Nodes which are offline and update the last_contact_failure.

That is not correct. There are 2 edge cases we would count a storage node as offline even if it is not.
1.) The download takes more than 25 seconds. As far as I know the audit job counts that as offline and will update the nodes table.
2.) The storage node wasn't pinged for a long time. In that case the audit job will assume that the node is offline and don't even try to contact it.

In general the audit job is the reason the old uptime tracking failed. The storage nodes are getting an uneven number of uptime checks and there is no ping involved. The audit job might flag you as offline even if you are available. I don't like the idea of running into the same trap again. Somewhere in the overlay service we have a job that is contacting the storage nodes every 2 hours. That sounds like a better job to identify offline nodes.

Reply:
The audit service only updates the `last_contact_failure`; there is no plan to reactivate the current uptime disqualification system, so the SN won't get disqualified.

The SN is only pinged by this new chore, audit or whatever else which currently updates the `last_contact_success` and `last_contact_failure` isn't going to be changed to do any ping.

Reply:
My plan is to add to the implementation section to remove the current uptime disqualification system because anyway, it's disabled.